lib/ez_gen/file_writer.rb

```diff
@@ -1,159 +1,66 @@
+require_relative 'template'
+
 class FileWriter
+  attr_reader :test_file
   def initialize(args)
-    @args = args
     @lower_name = args.last
     @upper_name = @lower_name.split('_').map(&:capitalize).join
+    @test_file = nil
```

Comment:
If the order isn't important, it might be good to just set `@test_file = write_test_file` in initialize, so there's no chance of calling the attr_reader returning `nil`

Reply:
Order does matter to some extent, none of these methods should get called until after `git init` has been run. I think it's actually helpful to have a default of `nil` here, it lets the developer know whether or not the test file has been created yet. Also, that's the point of making all of the `write_file` methods private, they should never be called from outside of the class... so if we set it in `initialize` that would effectively make the method public. One option we could do is just remove it from `initialize` altogether since the instance variable is actually being set from within the `write_test_file` method. Then if the attr_reader is called 'out of turn', the developer would get a `no method error`  instead of `nil` (although I think I still prefer `nil` as it won't immediately halt execution and is easier to look at the `FileWriter` class and discern what is happening).

Reply:
Currently we can call `write_file` which calls private methods. In terms of access this is equivalent to calling the same private method inside `initialize` once at instantiation.

I avoid `nil` instance variables because it requires calling methods in the correct order to get the expected return value--it feels like better abstraction/encapsulation when I don't have to know those details.

I have a better suggestion though. Why don't we move the `run_test` method inside the `FileWriter` class? It can be called at the end of `write`, so after all the files are written we immediately run tests. Two advantages:

1) We won't need `test_file` available outside `FileWriter` at all

2) We can probably capture the result of the tests and return it to `BasicEnv` as a success or failure of the `write` method. If the test fails, maybe we also choose not to commit and checkout a branch, and display a failure message with `MessageWriter`

Reply:
I implemented what I feel is an even simpler and better-encapsulated solution. I removed the instance variable `@test_file` and subsequent `attr_reader` from the `FileWriter` class in favor of just reading the file name in the `BasicEnv` class. Originally, I was just saving the test file name to a variable to make it easier to read/execute at the end of the script. But it feels like better encapsulation to me to not have the `FileWriter` class do any file reading/execution, that falls into the responsibilities of the `BasicEnv` class in my mind.